### PR TITLE
More balance tests

### PR DESCRIPTION
**Pull request by WyseNynja**
- [x] Refactors so tests use our objects (had to add Deserialize to some things)
- [x] include total_frontend_requests and total_cache_misses on Balance
- [x] figure out why stats are getting counted extra times and fix it

### CODE REVIEW

File: payment-contracts/src/contracts/ierc20.rs [hunk 1]
```diff
@@ -7,14 +7,16 @@ pub use ierc20::*;
     clippy::upper_case_acronyms,
     clippy::type_complexity,
     dead_code,
-    non_camel_case_types,
+    non_camel_case_types
 )]
 pub mod ierc20 {
     #[rustfmt::skip]
```
Comment (yenicelik) [inline payment-contracts/src/contracts/ierc20.rs:13]: should this contract be public?

File: payment-contracts/src/contracts/payment_factory.rs [hunk 2]
```diff
@@ -7,14 +7,16 @@ pub use payment_factory::*;
     clippy::upper_case_acronyms,
     clippy::type_complexity,
     dead_code,
-    non_camel_case_types,
```
Comment (yenicelik) [inline payment-contracts/src/contracts/payment_factory.rs:10]: should this be public?

File: web3_proxy/tests/test_proxy.rs [hunk 3]
```diff
@@ -13,15 +13,15 @@ use web3_proxy::rpcs::blockchain::ArcBlock;
 #[cfg_attr(not(feature = "tests-needing-docker"), ignore)]
 #[test_log::test(tokio::test)]
 async fn it_migrates_the_db() {
-    let x = TestApp::spawn(true).await;
+    let x = TestApp::spawn(31337, true).await;
```
Comment (yenicelik) [inline web3_proxy/tests/test_proxy.rs:16]: was it changed to 999_001_999 now? or is this number correct?

File: web3_proxy/tests/test_admins.rs [hunk 4]
```diff
@@ -66,6 +63,6 @@ async fn test_admin_grant_credits() {
 #[ignore = "under construction"]
 #[test_log::test(tokio::test)]
 async fn test_admin_change_user_tier() {
-    let x = TestApp::spawn(true).await;
+    let x = TestApp::spawn(31337, true).await;
```
Comment (yenicelik) [inline web3_proxy/tests/test_admins.rs:66]: same with 999_001_999

File: web3_proxy/tests/common/create_admin.rs [hunk 5]
```diff
@@ -38,15 +38,20 @@ pub async fn create_user_as_admin(
...
     info!(?admin_post_login_data);
 
-    let admin_login_response = r
+    let admin_post_login_data = r
         .post(&login_post_url)
         .json(&admin_post_login_data)
         .send()
         .await
         .unwrap()
-        .json::<LoginPostResponse>()
```
Comment (yenicelik) [inline web3_proxy/tests/common/create_admin.rs:47]: could directly read to serde_value prob (but no need to overcomplicate things)

File: web3_proxy/tests/common/app.rs [hunk 6]
```diff
@@ -244,7 +248,7 @@ impl TestApp {
         // TODO: test influx
         // TODO: test redis
         let app_config: AppConfig = serde_json::from_value(json!({
-            "chain_id": 31337,
+            "chain_id": chain_id,
```
Comment (yenicelik) [inline web3_proxy/tests/common/app.rs:251]: other test files use 31337 hardcoded

File: web3_proxy/tests/common/app.rs [hunk 7]
```diff
@@ -59,23 +60,26 @@ pub struct TestApp {
     pub proxy_provider: Provider<Http>,
 
     /// tell the app to flush stats to the database
-    flush_stat_buffer_sender: mpsc::Sender<oneshot::Sender<(usize, usize)>>,
+    flush_stat_buffer_sender: mpsc::Sender<oneshot::Sender<FlushedStats>>,
```
Comment (yenicelik) [inline web3_proxy/tests/common/app.rs:63]: idk what FlushedStats vs (usize, usize) is, but i suppose its better like this

File: web3_proxy/tests/test_sum_credits_used.rs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    // make one public rpc request of 16 CU
+    x.proxy_provider
+        .request::<_, Option<ArcBlock>>("eth_blockNumber", ("latest", false))
+        .await
+        .unwrap();
+
+    // make one authenticated rpc request of 16 CU
+    user_proxy_provider
+        .request::<_, Option<ArcBlock>>("eth_blockNumber", ("latest", false))
```
Comment (yenicelik) [inline web3_proxy/tests/test_sum_credits_used.rs:107]: will be cached just fyi

File: web3_proxy/tests/test_users.rs [hunk 9]
```diff
@@ -36,7 +36,7 @@ struct LoginPostResponse {
 #[cfg_attr(not(feature = "tests-needing-docker"), ignore)]
 #[test_log::test(tokio::test)]
 async fn test_log_in_and_out() {
-    let x = TestApp::spawn(true).await;
+    let x = TestApp::spawn(31337, true).await;
```
Comment (yenicelik) [inline web3_proxy/tests/test_users.rs:39]: again possible 999_001_999

File: web3_proxy/src/stats/mod.rs [hunk 10]
```diff
@@ -212,29 +224,23 @@ impl BufferedRpcQueryStats {
         self.sum_response_millis += stat.response_millis;
         self.sum_credits_used += stat.compute_unit_cost;
 
-        let latest_balance = stat.authorization.checks.latest_balance.read().await;
-        self.approximate_latest_balance_for_influx = latest_balance.clone();
+        if paid_credits_used {
```
Comment (yenicelik) [inline web3_proxy/src/stats/mod.rs:227]: where do we check if the user has a balance? ah its a function arg, i see

File: web3_proxy/src/stats/mod.rs [hunk 11]
```diff
@@ -372,59 +377,22 @@ impl BufferedRpcQueryStats {
...
 
-            // Apply all the referral logic; let's keep it simple and flat for now
             // Calculate if we are above the usage threshold, and apply a bonus
             // Optimally we would read this from the balance, but if we do it like this, we only have to lock a single table (much safer w.r.t. deadlocks)
             // referral_entity.credits_applied_for_referrer * (Decimal::from(10) checks (atomically using this table only), whether the user has brought in >$100 to the referer
             // In this case, the sender receives $100 as a bonus / gift
             // Apply a 10$ bonus onto the user, if the user has spent 100$
             // TODO: i think we do want a LockType::Update on this
             match referee::Entity::find()
+                .lock(LockType::Update)
```
Comment (yenicelik) [inline web3_proxy/src/stats/mod.rs:395]: why is this lock needed? we only do upserts, no?

File: web3_proxy/tests/test_proxy.rs (hunk 3 above)
Comment (WyseNynja) [reply web3_proxy/tests/test_proxy.rs:16]: 31337 is anvil's default network with ETH's CU price. 999_001_999 is just for having a very high CU cost to make tests simpler

File: web3_proxy/tests/common/create_admin.rs (hunk 5 above)
Comment (WyseNynja) [reply web3_proxy/tests/common/create_admin.rs:47]: reading to serde_value means a 10 line long thing thats half `unwrap` just to extract the actual types. using this was way simpler.

File: web3_proxy/tests/common/app.rs (hunk 6 above)
Comment (WyseNynja) [reply web3_proxy/tests/common/app.rs:251]: this is TestApp. It takes the chain_id from the other places

File: web3_proxy/src/stats/mod.rs (hunk 11 above)
Comment (WyseNynja) [reply web3_proxy/src/stats/mod.rs:395]: This one is fine:
```
                            referral_entity.one_time_bonus_applied_for_referee =
                                sea_orm::Set(bonus_for_user);
```

This one is not. This has a race:
```
                            referral_entity.credits_applied_for_referrer = sea_orm::Set(
                                referral_entity.credits_applied_for_referrer.as_ref()
                                    + referrer_bonus,
                            );
```

File: web3_proxy/src/stats/stat_buffer.rs [hunk 12]
```diff
@@ -127,22 +131,83 @@ impl StatBuffer {
...
+                                if user_id == 0 {
+                                    paid_credits_used = false;
+                                    latest_balance = Balance::default();
+                                } else {
+                                    // update the user's cached balance
+                                    let mut user_balance = stat.authorization.checks.latest_balance.write().await;
+
+                                    paid_credits_used = user_balance.active_premium();
+
+                                    user_balance.total_frontend_requests += 1;
```
Comment (yenicelik) [inline web3_proxy/src/stats/stat_buffer.rs:158]: this seems extremely manual and bug-prone

File: web3_proxy/src/stats/stat_buffer.rs [hunk 13]
```diff
@@ -127,22 +131,83 @@ impl StatBuffer {
...
+                                        // TODO: this lets them get a negative remaining balance. we should clear if close to 0
+                                        user_balance.total_spent_paid_credits += stat.compute_unit_cost;
+
+                                        // check if they still have premium
+                                        if user_balance.active_premium() {
+                                            // TODO: referall credits here? i think in the save_db section still makes sense for those
+                                        } else {
+                                            // was premium, but isn't anymore due to paying for this query. clear the cache
+                                            // TODO: stop at <$0.000001 instead of negative?
+                                            self.user_balance_cache.invalidate(&user_balance.user_id).await;
```
Comment (yenicelik) [inline web3_proxy/src/stats/stat_buffer.rs:174]: yes, doing some EPS makes sense

File: web3_proxy/src/stats/stat_buffer.rs [hunk 14]
```diff
@@ -127,22 +131,83 @@ impl StatBuffer {
...
+
+                                            let rpc_keys = rpc_key::Entity::find()
```
Comment (yenicelik) [inline web3_proxy/src/stats/stat_buffer.rs:176]: could this possibly not catch all the rpc keys? probably should be fine

**Comment by yenicelik** [comment]
checked everything lots of changes, added comments

File: web3_proxy/src/stats/stat_buffer.rs (hunk 12 above)
Comment (WyseNynja) [reply web3_proxy/src/stats/stat_buffer.rs:158]: theres only one place that has a .write now. but we could move it into a function on Balance

File: web3_proxy/src/stats/stat_buffer.rs (hunk 14 above)
Comment (WyseNynja) [reply web3_proxy/src/stats/stat_buffer.rs:176]: how would it miss one?

**Pull request merged**